src/pkcs15init/pkcs15-lib.c

```diff
@@ -784,8 +785,15 @@ sc_pkcs15init_add_app(struct sc_card *card, struct sc_profile *profile,
 	if (card->app_count >= SC_MAX_CARD_APPS)
 		LOG_TEST_RET(ctx, SC_ERROR_TOO_MANY_OBJECTS, "Too many applications on this card.");
 
+	/* In case of pinpad readers check if SO PIN is defined in a profile */
+	if (!args->so_pin_len && (card->reader->capabilities & SC_READER_CAP_PIN_PAD)) {
+		sc_profile_get_pin_info(profile, SC_PKCS15INIT_SO_PIN, &pin_ainfo);
+		/* If found, assume we want SO PIN */
+		has_so_pin = pin_ainfo.attrs.pin.reference != -1;
+	}
+
 	/* If the profile requires an SO PIN, check min/max length */
-	if (args->so_pin_len) {
+	if (args->so_pin_len || has_so_pin) {
```

Comment:
This looks like a hack: How can you check max/min length if `args->so_pin_len` is `0`? Shouldn't this look something like this:
```c
	/* In case of pinpad readers check if SO PIN is defined in a profile */
	if (!args->so_pin_len && (card->reader->capabilities & SC_READER_CAP_PIN_PAD)) {
		sc_profile_get_pin_info(profile, SC_PKCS15INIT_SO_PIN, &pin_ainfo);
		/* If found, assume we want SO PIN */
		has_so_pin = pin_ainfo.attrs.pin.reference != -1;
	} else {
		has_so_pin = args->so_pin_len != 0;
	}

	if (has_so_pin) {
		const char	*pin_label;

		/* If the profile requires an SO PIN, check min/max length */
		if (args->so_pin_len) {
			sc_profile_get_pin_info(profile, SC_PKCS15INIT_SO_PIN, &pin_ainfo);
			r = sc_pkcs15init_qualify_pin(card, "SO PIN", args->so_pin_len, &pin_ainfo);
			LOG_TEST_RET(ctx, r, "Failed to qualify SO PIN");
		}
```

Reply:
Well, since `sc_pkcs15init_qualify_pin()` returns `SC_SUCCESS` for `args->so_pin_len == 0`, I thought additional `if` wasn't necessary...

Anyway, here's an updated PR.